### ISSUE TITLE
[Feature Request]: Using serial port without access to the port

### ISSUE THREAD

**Issue by kdoelling1919** [psychopy/psychopy#6494]
### Description

The goal would be to add a feature to the Serial Port Out Component for testing purposes when you don't have access to the port in question. Ideally the code that would have been sent to the port is instead presented in a corner of the screen so that the script is unchanged with or without the port connected (with a noticeable warning saying the port was not found).

### Purpose

To test serial port behavior at times when the user doesn't have access to the port they will ultimately use. 

### Is your feature request related to a problem?

I use the serial port to send triggers to my EEG system, but if I'm working at home I don't have my EEG system on hand so the port doesn't work and I have to disable the component in order to test the rest of the script, nor can I try out different trigger patterns until I get to the EEG. Instead, I make a separate text component that writes the trigger in the top right corner with the same conditions as my serial port so that I can see if it works and then toggle back and forth between disabling various components when I move to the EEG room. Instead it would be easier and beneficial for the serial port component itself to have a testing feature which writes the trigger in the corner if the port is not found.

### Describe the solution you would like

Add a checkbox to the Serial Port Component to write trigger to screen if port not found. And a textbox for alternative text to display, as triggers are often sent in bytes format.
If the checkbox is checked, then the experiment should display a big bright warning screen at the beginning of the experiment if the port is not found.

### Describe alternatives that you have considered

Another way would be to have a checkbox which converts the serial to a testing mode so that it never tries to send to the port and only writes the text to the screen, (so the program doesn't automatically decide). I feel there is some risk here in terms of users forgetting to uncheck the box when they deploy the experiment in the lab but at least it's not the program's fault in that case. I think my git repo would more appreciate the solution I described though.

Another option would be for the toggle between writing to Port and writing to screen be controlled by the Pilot/Run toggle. I might like this option a bit better than the first alternate, but I can imagine the User might get annoyed trying to "Pilot" the triggers in the EEG room and not being able to have other Pilot features be true. I actually don't know if "Pilot" does anything yet for psychopy or if its related more to psychojs. 

### Additional context

I've not edited the Builder part of psychopy before, but if the PsychoPowers-that-be think this would be a relatively easy enhancement, I could take a stab at it.

**Comment by GordonDodwell** [comment]
This can be resolved by installing and configuring a serial port emulator on your system, e.g., com0com.

**Comment by peircej** [comment]
I like the idea in general @kdoelling1919 and maybe using com0com (which I'd never heard of, so thanks @GordonDodwell ) is a cool way to do it. We could use com0com to open up a little serial port terminal window, and that use the terminal window to send/receive rather than the physical device. The advantage of this over simply sending the strings to the screen is that means having the ability to send strings _back_ from the emulated device. Then again, that may create the new problem that the user may then need to type in all the expected device responses which would turn painful very quickly, so we'd need to have something so that serial devices in emulated mode didn't "expect" a response. 🤔 

All good food for thought

**Comment by GordonDodwell** [comment]
link to com0com for reference: https://sourceforge.net/projects/com0com/

@peircej The com0com solution essentially just emulates a persistent linked pair of aliased serial ports, so this wouldn't quite solve the entire request on its own (in terms of what Psychopy currently sees, the emulated "in" port appears no different than an active physical port). However, when a pair of ports is created using com0com, whatever data is sent by the experimental script to the "in" port would be directly output to the "out" port, which could then be piped into a text component (getting closer to the requested functionality).

A challenge may be having the experimental script dynamically identify what port(s) to use, as these could differ between a lab setup with physical devices and the emulated ports in the test environment. IMO it would be easiest to have an "emulation" mode on serial components that makes subprocess calls to com0com on initialisation, creating a temporary set of ports which override whatever port setting is in place in the script for physical hardware, then removes those emulated ports with a second call to com0com before exiting.

```python

def emulate_ports(action:str):
    """
    Function to emulate COM ports using com0com
    run './setupc.exe help' in the terminal for more info on passable parameters
    
    action: str, "add" or "remove"
    add: add a pair of emulated COM ports
    remove: remove the last pair of emulated COM ports
    """
    import subprocess

    com0com = "C:\\Program Files (x86)\\com0com\\setupc.exe"  # will need to be changed to whatever path of the setupc.exe file

    if action == "add":
        # get all used ports to avoid conflicts
        used_ports = subprocess.run(
            [com0com, "busynames COM*"], capture_output=True, text=True
        ).stdout.split("\n")[:-1]

        # get the last used port number
        last_port = max({int(x.split("COM")[1]) for x in used_ports})

        # generate names of new ports
        in_port_name = f"COM{last_port + 1}"
        out_port_name = f"COM{last_port + 2}"

        # set port names to the desired values
        in_port = f"PortName={in_port_name}"  # additional parameters can be added here, e.g., EmuBR / EmuOverrun
        out_port = f"PortName={out_port_name}"

        # generate emulated port pair
        subprocess.run([com0com, f"install {in_port} {out_port}"])

        # return the name of the generated in port
        return in_port_name

    elif action == "remove":
        # get all emulated ports (incase others are generated elsewhere)
        all_emulated_ports = subprocess.run(
            [com0com, "listfnames"], capture_output=True, text=True
        ).stdout

        # identify the reference of the last emulated pair
        last_emulated_pair = [
            int(x[-1]) for x in all_emulated_ports.split(" ") if "CNCB" in x
        ][-1]

        # remove the last emulated pair
        subprocess.run([com0com, f"remove {last_emulated_pair}"])

```

@kdoelling1919 for the mean time, if you just install com0com and create a paired set of ports with static names, you can then send whatever triggers you want to the "in" port, and the script **_should_** run without complaining that the port isn't available (I'm not 100% certain because I don't use builder, but its fairly straight forward in code). You would however need to use some code components to have the string being sent to the port (or alternatively, output to the second port) printed on screen.

**Comment by kdoelling1919** [comment]
I'm glad you find it appealing @peircej ! and Thanks @GordonDodwell for the code to show how com0com might work. Very informative.

My concern about using com0com (in terms of updating the builder component) would be adding another dependency for psychopy or requiring the user to install it separately. I'm not sure it's necessary for the builder to still actually use pyserial for the emulation mode for it to be useful. As the participant will still need to test  on the actual trigger port when comes time to run the experiment, I'm not sure greater realism provides much value. Also, as far as I can tell, com0com is only supported on Windows, is that right @GordonDodwell ?

@peircej's point about controlling responses of the port could be interesting but I've never needed to use this feature so I don't know how wide spread it is, but from my reading of the documentation, com0com only outputs what you input, so I'm not sure it could be used for such a feature. 

For now, I'm still Team "write the string to the screen" but I'm happy to see this idea move forward either way.

**Comment by GordonDodwell** [comment]
@kdoelling1919 for Linux and OSX, Socat can be used to the same effect as com0com, though would still require a new dependency be added. Netcat, a similar utility is usually preinstalled in OSX and most Linux distros, but I'm not sure if it has the same capabilities. 

Perhaps there are in effect two separate features to consider. The first being an option to have output trigger strings printed to screen on request (which would be useful for building / testing even when remote hardware is connected). The second being how to handle scenarios where the desired ports aren't available (e.g., home office). As you suggest, port emulation wouldn't be necessary at all if triggers aren't actually being sent - the serial component would just need to check whether the target port is available, and disable port communication if not. This is already possible with pyserial, so no need for new dependencies:

```python
import serial.tools.list_ports
ports = serial.tools.list_ports.comports()

for port, desc, hwid in sorted(ports):
        print("{}: {} [{}]".format(port, desc, hwid))
```

**Comment by kdoelling1919** [comment]
Fair enough, @GordonDodwell , maybe it is actually two different features. I'm a bit nervous about allowing the user to select the second feature by itself (Disabling port communication if the target port is not available) unless it comes with a super visible and annoying warning (and even those can be missed), so that people don't accidentally have the option clicked for a real experiment. 

One option would be to allow them to select feature 1 (print triggers to screen) regardless of target port being found, but forcing the trigger strings if the port is not found so that it remains visible that the port is not working.

**Comment by Cortexan** [comment]
@kdoelling1919  - agreed, I'd suggest that any disabling of port communication should be entirely handled in the background rather than giving users an option to select it. Perhaps this could be tied to a user-selectable "display port information" option, providing an on-screen overlay including information on the target port, whether or not its enabled, and the data being sent to it.

**Comment by GordonDodwell** [comment]
@kdoelling1919 - agreed, any disabling of port communication should be entirely handled in the background rather than giving users an option to select it. Perhaps this could be tied to a user-selectable "display port information" option, providing an on-screen overlay including information on the target port, whether or not its enabled, and the data being sent to it.

**Comment by peircej** [comment]
Definitely these are separate options - the print-to-screen is easier and would suit most use cases. I still think having a little GUI widget that can pretend to be another device is a useful/cool feature. To be honest, I don't see either option being a top priority in the short term (definitely don't expect these in 2024.2.0) but the print-to-screen option could be added potentially in the 2025.1 series

**Comment by peircej** [comment]
> My concern about using com0com (in terms of updating the builder component) would be adding another dependency for psychopy or requiring the user to install it separately. I'm not sure it's necessary for the builder to still actually use pyserial for the emulation mode for it to be useful. As the participant will still need to test on the actual trigger port when comes time to run the experiment, I'm not sure greater realism provides much value. Also, as far as I can tell, com0com is only supported on Windows, is that right @GordonDodwell ?

Now that we have plugins the emulation option would be something to add as a plugin I think
